drivers: rtc: Add RV3028 RTC driver

jakubtopic opened the pull request:
This PR adds basic support for [microcrystal RV3028](https://www.microcrystal.com/fileadmin/Media/Products/RTC/App.Manual/RV-3028-C7_App-Manual.pdf) RTC connected to the I2C bus:
- It supports getting/setting time and time alarm/update callbacks.
- Backup switchover and trickle charger configuration using devicetree.
- ~~I also added an API extension for accessing the user RAM registers and EEPROM.~~
  - Accessing the user RAM and EEPROM will be handled using the retained_mem API in a separate PR
- I didn't implement the device PM API because it wasn't clear to me how it should behave in terms of power switchover functionality.

I have tested the driver using the RTC API test suite. The Y2K test is skipped because this RTC only supports the final two digits of years.

Thanks

jakubtopic commented:
Oh, I just realized my PR overlaps with #73385. I didn't see it at the time of submission. This is my first zephyr contribution, so I'd still appreciate any suggestions for the next time 🙂

bjarki-trackunit commented:
Wild timing, only 1 hour after #73385 was pushed :) It would be great if you help review #73385 since you probably have quite a bit of knowledge of that specific RTC :)

decsny commented on dts/bindings/rtc/microcrystal,rv3028.yaml:
```diff
@@ -0,0 +1,73 @@
+        switchover occurs from VDD to VBACKUP
+
+  aux-voltage-chargeable:
+    type: int
```
should this be a boolean?

jakubtopic replied:
it could definitely be a boolean, but I thought the intent was to have DT bindings similar to Linux where this property is an integer with enum values

decsny replied:
It would be helpful to have a link or something then. It looks like this property is generic to RTC in Linux.

jakubtopic replied:
here is the generic RTC binding in the Linus's tree: [https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/rtc/rtc.yaml](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/rtc/rtc.yaml)

jakubtopic replied:
Here is the existing binding in the Linus's tree:
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/rtc/microcrystal%2Crv3028.yaml

And the generic RTC binding:
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/rtc/rtc.yaml

jakubtopic replied:
@decsny should I change the datatype or leave it like the existing Linux bindings?

jakubtopic replied:
ok, sounds good :)

jakubtopic replied:
@bjarki-trackunit I’ve ended up removing the property `aux-voltage-chargeable` because I felt it was redundant:
https://github.com/zephyrproject-rtos/zephyr/blob/5f8769677117be5a5262b7903c05b261e4c56e9d/dts/bindings/rtc/microcrystal%2Crv3028.yaml#L40-L54

Kampi commented:
Hi @jakubtopic and @bjarki-trackunit,

I removed the RV-3028 stuff from my PR because the driver from @jakubtopic is more advanced than my driver. I will focus on RV-8263 because this RTC is more important to me. 👍

bjarki-trackunit commented on drivers/rtc/Kconfig.rv3028:
```diff
@@ -0,0 +1,30 @@
+
+if RTC_ALARM || RTC_UPDATE
+
+config RTC_RV3028_THREAD_STACK_SIZE
+	int "Stack size for the RV3028 interrupt thread"
+	default 512
+	help
+	  Size of the stack used for the thread handling interrupts and dispatching callbacks.
+
+config RTC_RV3028_THREAD_PRIO
+	int "Priority for the RV3028 interrupt thread"
+	default 0
+	help
+	  Priority level for the thread handling interrupts and dispatching callbacks.
```
You really should be using the system workqueue for handling callbacks like this, a dedicated thread for handling a single callback is really expensive, see https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/include/zephyr/kernel.h#L3356

jakubtopic replied:
It was indeed an oversight on my part :) Thanks for your feedback and for pointing me to the relevant source.

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,1101 @@
+
+	if (timeptr == NULL ||
+	    (timeptr->tm_year < RV3028_YEAR_OFFSET) ||
+	    (timeptr->tm_year > (RV3028_YEAR_OFFSET + 99))) {
```
You may want to additionally validate the time using https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/drivers/rtc/rtc_utils.h#L26

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,1101 @@
+		LOG_ERR("Unsupported alarm field mask 0x%04x", mask);
+		return -EINVAL;
+	}
+
```
Please validate the timeptr using https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/drivers/rtc/rtc_utils.h#L26 here as well

jakubtopic replied:
I've updated the time validation

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,1101 @@
+	if (err) {
+		return err;
+	}
+
```
If the time is invalid, this function should return -ENODATA, for the RV3028, the POR (power-on-reset) flag should be able to determine that the time is not valid :) There may be a better way to check but POR flag seems perfect :)

jakubtopic replied:
Done

bjarki-andreasen reviewed (changes requested):
Regarding the Y2K test failing, it really shouldn't be, if the year can't be set to 1999, the RTC should return -EINVAL, and the test should be skipped, see https://github.com/zephyrproject-rtos/zephyr/blob/3570408db556fbb53cd775ae3d2e44553e28cb31/tests/drivers/rtc/rtc_api/src/test_y2k.c#L44-L51
It may be a legitimate bug :)

Kampi commented on include/zephyr/drivers/rtc/rtc_rv3028.h:
```diff
@@ -0,0 +1,84 @@
+ * @retval -EINVAL Invalid address parameter.
+ * @retval -EIO General input / output error.
+ */
+__syscall int rv3028_ram_read(const struct device *dev, uint8_t addr, uint8_t *val);
```
@bjarki-trackunit does it make sense to make a generic function out of it? Because some RTCs (like the RV-8263 and the DS1307) have internal RAM which can be used.

jakubtopic replied:
Yes, that was my question too. Currently it is not possible for one device to implement multiple APIs, right?

Kampi replied:
It's not supported by the RTC API. The RTC API can not use a second API (same for other device types) do you can not combine them. What you need are two functions in the API structure for reading and writing RAM.

bjarki-trackunit replied:
The solution is to expose the memory as a memory device, which is a child of the RTC. The memory device will then implement an appropriate API for memory access (may I suggest https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/retained_mem.h)

An example devicetree layout
```
rtc: rtc@2 {
        compatible: "microcrystal,rv3028";
        status = "okay";

        retmem: retmem {
                compatible: "microcrystal,rv3028-retmem";
                status = "okay";
        };
};
```
The rtc node implements the RTC API, the retmem node implements the retained_mem API. The device driver for the rv3028 defines both devices, both devices share the same data and config structures :)

This may be a bit overly complicated for this PR, I would add this in the future

jakubtopic replied:
Ah this is a bummer 😕 I would prefer to be able to use the retention subsystem API or something like that. This does not only apply to RTC – many other devices contain retained registers. I think the RTC API should be limited to the time stuff.

bjarki-trackunit replied:
This problem is way larger than just RTCs, think of modems that contain a GNSS, cellular radio, SMS, and often whole flash filesystems, representing the "internal" hardware blocks of devices and modems as individual devices seems to be the only scalable way to fully expose the features of any device.

RTCs often contain watchdogs and low power counters as well, we can't support all of this in one API :)

Kampi replied:
So we have to add an additional retmem driver to support this?

jakubtopic commented:
Hi @Kampi, I wouldn't say it is more advanced, but I appreciate your trust. I'll keep working on my driver then. Good luck with the other device!

jakubtopic committed:
dts: bindings: add binding for the RV3028 RTC

Adds a devicetree binding for the Micro Crystal RV3028 RTC connected to
the I2C bus.

Signed-off-by: Jakub Topic <jakub.topic@anitra.cz>

jakubtopic committed:
drivers: rtc: add RV3028 RTC driver

Adds support for the Micro Crystal RV3028 RTC connected to the I2C bus.

Signed-off-by: Jakub Topic <jakub.topic@anitra.cz>

jakubtopic commented:
- I was actually running an older version of the Y2K test. It looks ok now
- removed the `rtc_rv3028.h` header file  look into this later
- replaced the dedicated thread with the system workqueue
- added time validation
- the backup registry value is determined directly in the device instance

jakubtopic commented:
@Kampi, do you want me to implement the RTC calibration?

jakubtopic commented:
I would suggest removing this test case dependency so we can test boards that have RTC connected externally but don't have it listed in the supported features.
https://github.com/zephyrproject-rtos/zephyr/blob/362588037efcbbb50557e5e3cb19d799165bf5c6/tests/drivers/rtc/rtc_api/testcase.yaml#L11

Kampi commented:
> @Kampi, do you want me to implement the RTC calibration?

The RTC doesn´t have a calibration function so you have to add it (somehow) in the driver. Not sure if this would make sense.

jakubtopic commented:
By _calibration_ I meant setting frequency offset correction using the `rtc_set_calibration` function. I noticed it in your driver for RV8263, so I wasn't sure if you needed it for RV3028 as well.

Kampi commented:
> By _calibration_ I meant setting frequency offset correction using the `rtc_set_calibration` function. I noticed it in your driver for RV8263, so I wasn't sure if you needed it for RV3028 as well.

I added it because the RTC contains a register for a calibration value so it's done in hardware. You can do a software calibration for the RV3028 too and use this interface.

jakubtopic commented:
> I added it because the RTC contains a register for a calibration value so it's done in hardware. You can do a software calibration for the RV3028 too and use this interface.

I think both ICs support frequency offset compensations. It's not a feature I need, so I was wondering if you hadn't originally planned it for RV3028 as well 🙂

Kampi commented:
> > I added it because the RTC contains a register for a calibration value so it's done in hardware. You can do a software calibration for the RV3028 too and use this interface.
> 
> I think both ICs support frequency offset compensations. It's not a feature I need, so I was wondering if you hadn't originally planned it for RV3028 as well 🙂

No I haven´t planed it for RV3028 because I haven´t found a calibration register or function during a quick check. :)

jakubtopic commented:
Fixed formatting errors

jakubtopic commented:
@bjarki-trackunit, @decsny Kindly asking for re-review :)

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,980 @@
+	}
+
+	err = rv3028_read_reg8(dev, RV3028_REG_STATUS, &status);
+	if (err) {
+		return err;
+	}
+
+	if ((status & RV3028_STATUS_PORF) != 0) {
```
Just use `if (status & RV3028_STATUS_PORF) {` like you do for `if (err) {` :)

jakubtopic replied:
I've simplified the other instances of this explicit comparison as well. I wasn't sure if this was an acceptable code style :)

bjarki-trackunit commented on drivers/rtc/rtc_rv3028.c:
```diff
@@ -0,0 +1,980 @@
+};
+
+struct rv3028_data {
+	struct k_mutex lock;
```
Please use a `k_sem` for locking if the owner of the lock is not used (like it would be for a condvar or recursive locking)

Like this:
https://github.com/zephyrproject-rtos/zephyr/blob/593dfe13856df4d2919874037cbb5691b9f00a15/drivers/gnss/gnss_emul.c#L49-L61
and
https://github.com/zephyrproject-rtos/zephyr/blob/593dfe13856df4d2919874037cbb5691b9f00a15/drivers/gnss/gnss_emul.c#L470

jakubtopic replied:
In terms of semantics, a mutex seemed like a more intuitive choice, but I honestly had no idea it was more expensive than semaphores 😊

bjarki-andreasen reviewed (changes requested):
The driver looks really good, Minor nits and a request to use k_sem for locking to optimize size and speed compared to a mutex :)

jakubtopic reviewed (commented):
@bjarki-trackunit Thanks for all the feedback. I've addressed the requested changes.

decsny commented on dts/bindings/rtc/microcrystal,rv3028.yaml:
```diff
@@ -0,0 +1,64 @@
+      - 32
+      - 1
+    description: |
+      Frequency of the CLKOUT signal in Hertz (Hz). When disabled, the CLKOUT pin is LOW.
```
what does "disabled" mean here? I think it should be clarified

jakubtopic replied:
I'll update the description to say: `If omitted, the CLKOUT pin is LOW.` Is it ok?

decsny commented on dts/bindings/rtc/microcrystal,rv3028.yaml:
```diff
@@ -0,0 +1,64 @@
+      - 15000
+    description: |
+      Resistance selection for the series current limiting resistor of the trickle charger:
+        - 3000: 3 kOhm (default value on delivery)
```
there should be a short sentence saying why 3000 is the default.

jakubtopic replied:
Thanks for pointing this out. I used `default:` because it matches the vendor's provided default value. Should I make this property required instead? It also depends on `aux-voltage-chargeable` being true.

jakubtopic replied:
@decsny To avoid incorrect configuration, I've removed the property `aux-voltage-chargeable`, which is redundant. Enabling the trickle charger is done by specifying `trickle-resistor-ohms`.

decsny reviewed (commented):
binding seems fine

bjarki-trackunit commented on dts/bindings/rtc/microcrystal,rv3028.yaml:
```diff
@@ -0,0 +1,59 @@
+    description: |
+      Resistance Selection for the Trickle Charger
+
+      This property specifies the resistance value for the series current-limiting resistor in the
+      trickle charging circuit. It should be provided if the trickle charger is to be enabled. The
+      trickle charger allows a battery or supercapacitor connected to VBACKUP to be charged.
```
I really like the new this binding, but the description could be a bit simplified:
```
Enable the trickle charger with provided current-limiting resistance in ohms. The trickle charger allows a battery or supercapacitor connected to VBACKUP to be charged.
```
Specifically, `It should be provided if the trickle charger is to be enabled` sounds like there is something more that needs to be done to actually enable trickle charging, which is disputed on line 54

jakubtopic replied:
Thanks for the suggestion. Done.

bjarki-andreasen reviewed (dismissed):
Great! I have one suggestion but it is just a minor nit IMO, great work!
